File: p8modem.lua
--
-- p8modem.lua
--
-- "Firmware" for p8modem, the virtual PICO-8 modem.
-- #include this file in your own PICO-8 programs in order
-- to use the p8modem API to send/receive packets over WebSockets.
-- See the documentation for more information.
--
-- (c) 2022 Joshua Moore
-- This program is free software, licensed under the terms
-- of the GNU GPLv3. See LICENSE.txt for details.
--


-- address of the first byte of PICO-8's 128-byte GPIO
-- memory space
-- the GPIO memory is used to transfer packet data between
-- the JS "hardware" and Lua "firmware"
_gpio_base = 0x5f80


-- transmit a packet to the websocket modem over the GPIO interface
-- buf should be a byte array or stringcontaining the desired tx packet data
--   buf should be at most 127 bytes long
--   bytes after the 127th will be ignored
-- returns the actual number of bytes sent
function p8m_send(buf)
  -- write 1 to GPIO[0] to signal that an outgoing packet
  -- is being prepared
  poke(_gpio_base, 1)

  -- write packet data to GPIO memory
  txcount = 0
  for i = 1, #buf do
    -- packet data is limited to 127 bytes
    if i > 127 then
      break
    end

    -- handle both strings and byte arrays
    if type(buf) == "string" then
      -- unpack string into GPIO memory
      poke(_gpio_base + i, ord(sub(buf, i, i)) & 0xff)
    elseif type(buf) == "table" then
      -- unpack byte array into GPIO memory
      poke(_gpio_base + i, buf[i] & 0xff)
    -- else
    -- unsupported type - silently ignore packet
    end

    txcount = txcount + 1
  end

  -- increment once more to account for packet length byte
  txcount = txcount + 1

  -- finally, write total packet length to GPIO[0]
  -- this signals to the modem that there is a packet ready to send
  poke(_gpio_base, txcount)

  return txcount
end


-- read a packet from GPIO memory, returning a byte array
-- containing only the packet data (without the leading packet
-- length byte)
-- returns nil if there is no rx packet waiting to be read
function p8m_recv()
  plen = peek(_gpio_base)

  -- if there is an unsent tx packet in GPIO memory,
  -- do not overwrite it with an rx packet
  if plen < 128 then
    return nil
  end

  -- convert signed negative length to positive value
  plen = (128 - plen) & 0x7f

  -- deserialize incoming packet
  packet = { }
  for i = 1, plen - 1 do
    add(packet, peek(_gpio_base + i))
  end

  -- write 0 to to GPIO[0]
  -- this signals that we are done reading the incoming packet,
  -- freeing GPIO memory
  poke(_gpio_base, 0)

  return packet
end

File: p8modem.js
//
// p8modem.js
//
// "Virtual modem" implementation of p8modem.
// This module acts as the bridge between the p8modem's Lua
// firmware API and the underlying WebSocket connection.
// The Lua API cooperates with the "virtual hardware" implementation
// in this file for the purposes of transferring data between the
// PICO-8 and the JS context across the GPIO memory interface.
//
// To use p8modem, export your PICO-8 program as a web application,
// then add a script tag to the HTML file which uses this script
// as its src. Then change SOCKET_URL below to point to the desired
// WebSocket URL.
// Once you've done that, you will be able to send and receive
// packets to and from the WebSocket server by including the Lua API
// from "p8modem.lua".
// See the documentation for more information.
//
// (c) 2022, Joshua Moore
// This program is free software, licensed under the terms
// of the GNU GPLv3. See LICENSE.txt for details.
//

// replace this with the WebSocket URL you want to connect to
const SOCKET_URL = "ws://localhost:8888";

// incoming packet queue
const _p8m_queue = new Array();


// pop a packet from the rx queue and write it to GPIO memory
const p8m_rx = () => {
  // receive incoming packet
  const packet = _p8m_queue.shift();

  if (packet !== undefined) {
    // write -1 to GPIO[0], indicating that we are processing an rx packet
    pico8_gpio[0] = 255;

    // convert packet length to a negative signed byte
    const plen = packet[0];
    const plen_signed = (128 - plen) | 0x80;

    for (let i = 1; i < plen; i++) {
        pico8_gpio[i] = packet[i];
    }

    // write negative packet length to GPIO[0]
    // this signals to the PICO-8 that there is an incoming
    // packet available to be read
    pico8_gpio[0] = plen_signed;
  }
};


// read a packet from GPIO memory and send it to the server
const p8m_tx = () => {
  // read outgoing packet
  const tx_packet = new Array();
  const plen = pico8_gpio[0];
  tx_packet.push(plen);
  for (let i = 1; i < plen; i++) {
    tx_packet.push(pico8_gpio[i]);
  }

  // send packet array in JSON array format
  socket.send("[" + tx_packet + "]");

  // finally write 0 to GPIO[0],
  // freeing GPIO memory for further use
  pico8_gpio[0] = 0;
};


// handle incoming/outgoing packets
const p8m_update = () => {

  if (pico8_gpio[0] !== 0) {
    // data is present on GPIO memory
    const plen = pico8_gpio[0];

    if (plen > 1 && plen < 128) {
      // send outgoing packet
      p8m_tx();
    }
  }

  if (!pico8_gpio[0]) {
    // GPIO memory is free, receive next incoming packet
    p8m_rx();
  }
};


// prepare websocket
const socket = new WebSocket(SOCKET_URL);

socket.onopen = (e) => {
  setInterval(p8m_update, 10);
};

socket.onmessage = (e) => {
  const rx_packet = JSON.parse(e.data);
  _p8m_queue.push(rx_packet);
};

socket.onclose = (e) => {
  alert("p8modem has lost connection");
};
